src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionEndpoint.cs

```diff
@@ -108,25 +142,89 @@ private async Task<IEnumerable<RazorCodeAction>> GetCodeActionsAsync(CodeActionP
                 return null;
             }
 
+            var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
+
+            var context = new RazorCodeActionContext(
+                request,
+                documentSnapshot,
+                codeDocument,
+                sourceText,
+                _languageServerFeatureOptions.SupportsFileManipulation);
+
+            return context;
+        }
+
+        private async Task<IEnumerable<RazorCodeAction>> GetCSharpCodeActionsAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+        {
+            var csharpCodeActions = await GetCSharpCodeActionsFromLanguageServerAsync(context, cancellationToken);
+            var filteredCSharpCodeActions = await FilterCSharpCodeActionsAsync(context, csharpCodeActions, cancellationToken);
+
+            return filteredCSharpCodeActions;
+        }
+
+        private async Task<IEnumerable<RazorCodeAction>> FilterCSharpCodeActionsAsync(RazorCodeActionContext context, IEnumerable<RazorCodeAction> csharpCodeActions, CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+               return null;
+            }
+
+            var tasks = new List<Task<RazorCodeAction[]>>();
+
+            foreach (var provider in _csharpCodeActionProviders)
+            {
+                var result = provider.ProvideAsync(context, csharpCodeActions, cancellationToken);
+                if (result != null)
+                {
+                    tasks.Add(result);
+                }
+            }
+
+            return await ConsolidateCodeActionsFromProvidersAsync(tasks, cancellationToken);
+        }
+
+        private async Task<IEnumerable<RazorCodeAction>> GetCSharpCodeActionsFromLanguageServerAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+        {
+            Range projectedRange = null;
+            if (context.Request.Range != null &&
+                !_documentMappingService.TryMapToProjectedDocumentRange(
+                    context.CodeDocument,
+                    context.Request.Range,
+                    out projectedRange))
+            {
+                return Array.Empty<RazorCodeAction>();
+            }
+
+            context.Request.Range = projectedRange;
+
             if (cancellationToken.IsCancellationRequested)
```

Comment:
I guess it was like this in other places too. I'm curious why.

Reply:
https://github.com/dotnet/aspnetcore-tooling/pull/2501#discussion_r489011372